found that the message is not safe to write

test code for reproduce:
```
client := NewMqClient(ub, auth)
client.df = JsonDataFormater{}
sr := loadServerP12Cert("obFm2pTw1kT8ADG7")
client.SetSecureRequst(envTestSrKid, sr)
obj := testObj{}
client.SubscribeRaw("****/p2p/consumerId@@@server09", obj)
num := uint32(0)
sem := semaphore.NewWeighted(80)
ctx := context.Background()
for {
  if err := sem.Acquire(ctx, 1); err != nil {
    t.Error("sem err,", err)
  }
  go func() {
    msgId := atomic.AddUint32(&num, 1)
    src := []byte(fmt.Sprintf(`{"msgType":"get","fromId":"server09","fromType":2,"msgId":%d,"msgDate":1637635559,"services":{"pluginsharealbum_create":{"share_key":"90a87a21785c4af28bd07fc76d077264","url":"http://127.0.0.1:8070?shareKey=90a87a21785c4af28bd07fc76d077264","name":"9","cover_path":"/6B653C174CCFAB6DD62A73DC3FD594C9/.HOMES/18665875540/00/IMG_20210812_112814.jpg","kid":"3033417224"}},"routePath":"cloud_interal"}`, msgId))
    client.PublishRaw("****/p2p/consumerId@@@hy_noid_000000000056/agent", nil, src)
    sem.Release(1)
  }()
}
```
```
 >20 Concurrent publish message, client disconnected
```

server requested disconnect (reason: 129)
server requested disconnect: &{ServerReference: ReasonString: SessionExpiryInterval:<nil> User:[]}
connection to broker lost (server requested disconnect (reason: 129)); will reconnect
Problems caused by NOACK

[warning] server09@****:7266 [Channel] The PUBACK PacketId 12987 is not found.

It appears that the mqtt broker repeatedly sends the same message to the client, and there has been no ack, causing the server to disconnect the connection

server error message:
```
2021-11-24T11:34:07.411614+08:00 [error] deviceAgent-hy_noid_000000000056@120.234.47.82:52542 [MQTT] , Parse failed for function_clause, [{emqx_frame,parse_utf8_string,[<<50,201,1,0,36,100,109,95,97,105,114,100,105,115,107,47,112,50,112,47,99,111,110,115,117,109,101,114,73,100,64,64,64,115,101,114,118,101,114,48,57,0,12,2,1,0,123,34,101,114,114,111,114,95,99,111,100,101,34,58,48,44,34,109,115,103,84,121,112,101,34,58,34,103,101,116,82,101,115,112,34,44,34,102,114,111,109,73,100,34,58,34,104,121,...>>],[{file,"emqx_frame.erl"},{line,464}]},{emqx_frame,parse_packet,3,[{file,"emqx_frame.erl"},{line,249}]},{emqx_frame,parse_frame,4,[{file,"emqx_frame.erl"},{line,183}]},{emqx_connection,parse_incoming,3,[{file,"emqx_connection.erl"},{line,625}]},{emqx_connection,handle_msg,2,[{file,"emqx_connection.erl"},{line,618}]},{emqx_connection,process_msg,2,[{file,"emqx_connection.erl"},{line,364}]},{emqx_connection,handle_recv,3,[{file,"emqx_connection.erl"},{line,328}]},{proc_lib,wake_up,3,[{file,"proc_lib.erl"},{line,236}]}], Frame data:<<50,201,1,50,201,1,0,36,100,109,95,97,105,114,100,105,115,107,47,112,50,112,47,99,111,110,115,117,109,101,114,73,100,64,64,64,115,101,114,118,101,114,48,57,0,12,2,1,0,123,34,101,114,114,111,114,95,99,111,100,101,34,58,48,44,34,109,115,103,84,121,112,101,34,58,34,103,101,116,82,101,115,112,34,44,34,102,114,111,109,73,100,34,58,34,104,121,95,110,...>>
2021-11-24T11:34:08.013642+08:00 [error] deviceAgent-hy_noid_000000000056@120.234.47.82:52548 [MQTT] , Parse failed for function_clause, [{emqx_frame,parse_utf8_string,[<<"\"services\":{\"pluginsharealbum_create\":null}}">>],[{file,"emqx_frame.erl"},{line,464}]},{emqx_frame,parse_packet,3,[{file,"emqx_frame.erl"},{line,249}]},{emqx_frame,parse_frame,4,[{file,"emqx_frame.erl"},{line,183}]},{emqx_connection,parse_incoming,3,[{file,"emqx_connection.erl"},{line,625}]},{emqx_connection,handle_msg,2,[{file,"emqx_connection.erl"},{line,618}]},{emqx_connection,process_msg,2,[{file,"emqx_connection.erl"},{line,364}]},{emqx_connection,handle_recv,3,[{file,"emqx_connection.erl"},{line,328}]},{proc_lib,wake_up,3,[{file,"proc_lib.erl"},{line,236}]}], Frame data:<<"7,\"services\":{\"pluginsharealbum_create\":null}}{\"error_code\":0,\"msgType\":\"getResp\",\"fromId\":\"hy_noid_000000000056\",\"fromType\":0,\"msgId\":840,\"msgDate\":1637724847,\"services\":{\"pluginsharealbum_create\":null}}2�"...>>
```

found message data sticky packets, Frame data error.

![image](https://user-images.githubusercontent.com/17872894/143183795-56aaf8ba-d8d9-4205-a33a-50e9d5b42bf0.png)


OR 
```
// Packets WriterTo
whatWriteTo struct {
  w io.WriterTo
  e error
}

Client struct {
  mu sync.Mutex
  ClientConfig
  // raCtx is used for handling the MQTTv5 authentication exchange.
  raCtx          *CPContext
  stop           chan struct{}
  publishPackets chan *packets.Publish
  acksTracker    acksTracker
  workers        sync.WaitGroup
  serverProps    CommsProperties
  clientProps    CommsProperties
  serverInflight *semaphore.Weighted
  clientInflight *semaphore.Weighted
  debug          Logger
  errors         Logger
  outPackets     chan *whatWriteTo
}
```
newClient:
```
// new zero size chan for block
&Client{
  outPackets: make(chan *whatWriteTo),
}
go waitWrite()
```

add sub func:
```
func (c *Client) waitWrite() {
	for {
		wwt, ok := <-c.outPackets
		if !ok {
			return
		}
		_, wwt.e = wwt.w.WriteTo(c.Conn)
	}
}

func (c *Client) doWrite(sp io.WriterTo) error {
	wt := &whatWriteTo{
		w: sp,
	}
	c.outPackets <- wt
	return wt.e
}
```
Client object  sub function `*.WriteTo(c.Conn)`  to  `c.doWrite(*)`

Hi, thank you for the prompt response, note that I'm not the original issue submitter.

Given its usage, `FixedHeader` seems ok, seems I extrapolated the issue a bit too far, sorry about that.

There is a race condition when writing `ControlPacket`s to `net.Conn` because there is no control over who's writing. For example, if two goroutines publish at the same time, their ControlPackets might get mixed up. I noticed this issue in a resource-constrained environment when I somehow published a message at the same time as PingHandler made a `PINGREQ`. Since there's strict packet validation on the server, my client was disconnected immediately.

A test that reproduces the issue: extended `packets/packets_test.go` with two benchmarks, the only difference is `sync.Mutex` usage in the second one, which is in line with the second proposed solution from my previous comment. The reader goroutine reads packets and there are 10 goroutines writing to `io.Pipe`. If the writes are not coordinated with a lock, they become mixed up. This is because `net.Buffers` is a `[][]byte` and its `WriteTo` method calls `io.Write` in a for loop - when `io.Write` is not being executed, this is an opportunity for another goroutine to start writing its packet.

```go
func BenchmarkWriteTo(b *testing.B) {
	x := NewControlPacket(PUBLISH)
	x.Content.(*Publish).QoS = 0
	x.Content.(*Publish).Topic = "testTopic"
	x.Content.(*Publish).PacketID = uint16(100)
	x.Content.(*Publish).Payload = []byte("testPayload")
	pp := x.Content.(*Publish)

	r, w := io.Pipe()

	var wg sync.WaitGroup

	done := make(chan int, 1)
	go func(r io.Reader, done chan int) {
		for {
			_, err := ReadPacket(r)
			if err != nil {
				b.Error(err)
			}

			select {
			case <-done:
				return
			default:
			}
		}
	}(r, done)

	for i := 0; i < 10; i++ {
		wg.Add(1)
		go func() {
			for n := 0; n < b.N; n++ {
				pp.WriteTo(w)
			}
			wg.Done()
		}()
	}

	wg.Wait()
	done <- 1
}

func BenchmarkWriteTo_Mutex(b *testing.B) {
	x := NewControlPacket(PUBLISH)
	x.Content.(*Publish).QoS = 0
	x.Content.(*Publish).Topic = "testTopic"
	x.Content.(*Publish).PacketID = uint16(100)
	x.Content.(*Publish).Payload = []byte("testPayload")
	pp := x.Content.(*Publish)

	r, w := io.Pipe()

	var wg sync.WaitGroup
	done := make(chan int, 1)

	go func(r io.Reader, done chan int) {
		for {
			_, err := ReadPacket(r)
			if err != nil {
				b.Error(err)
			}

			select {
			case <-done:
				return
			default:
			}
		}
	}(r, done)

	var mu sync.Mutex
	for i := 0; i < 10; i++ {
		wg.Add(1)
		go func() {
			for n := 0; n < b.N; n++ {
				mu.Lock()
				pp.WriteTo(w)
				mu.Unlock()
			}
			wg.Done()
		}()
	}

	wg.Wait()
	done <- 1
}
```

Gives the following output (shortened):
```
[~/git/paho.golang/packets] go test -bench 'BenchmarkWriteTo(_Locked)?'
goos: darwin
goarch: arm64
pkg: github.com/eclipse/paho.golang/packets
BenchmarkWriteTo-8              --- FAIL: BenchmarkWriteTo-8
    packets_test.go:445: EOF
    packets_test.go:445: invalid Prop type 0 for packet 4
    packets_test.go:445: invalid Prop type 101 for packet 4
    packets_test.go:445: invalid Prop type 116 for packet 4
    packets_test.go:445: unknown packet type 0 requested
.
.
.
    packets_test.go:445: invalid Prop type 105 for packet 4
    packets_test.go:445: invalid Prop type 115 for packet 4
    packets_test.go:445: invalid Prop type 48 for packet 4
    packets_test.go:445: unknown packet type 0 requested
    packets_test.go:445: invalid Prop type 112 for packet 4
    packets_test.go:445: invalid Prop type 48 for packet 4
    packets_test.go:445: invalid Prop type 48 for packet 4
    packets_test.go:445: unknown packet type 0 requested
    packets_test.go:445: invalid Prop type 112 for packet 4
    packets_test.go:445: invalid Prop type 0 for packet 4
    packets_test.go:445: invalid Prop type 23 for packet 4
BenchmarkWriteTo_Mutex-8           27843             42121 ns/op
FAIL
exit status 1
FAIL    github.com/eclipse/paho.golang/packets  1.788s
```

--- a/autopaho/net.go
+++ b/autopaho/net.go
@@ -14,6 +14,7 @@ import (
 
 	"github.com/gorilla/websocket"
 
+	"github.com/eclipse/paho.golang/packets"
 	"github.com/eclipse/paho.golang/paho"
 )
 
@@ -86,7 +87,8 @@ func attemptTLSConnection(ctx context.Context, tlsCfg *tls.Config, address strin
 	d := tls.Dialer{
 		Config: tlsCfg,
 	}
-	return d.DialContext(ctx, "tcp", address)
+	conn, err := d.DialContext(ctx, "tcp", address)
+	return packets.NewThreadSafeConn(conn), err
 }
 
 // attemptWebsocketConnection - makes a single attempt at establishing a websocket connection with the broker
@@ -114,7 +116,8 @@ func attemptWebsocketConnection(ctx context.Context, tlsc *tls.Config, cfg *WebS
 	}
 
 	wrapper := &websocketConnector{
-		Conn: ws,
+		Conn:   ws,
+		Locker: &sync.Mutex{},
 	}
 	return wrapper, err
 }
@@ -126,7 +129,9 @@ type websocketConnector struct {
 	*websocket.Conn
 	r   io.Reader
 	rio sync.Mutex
-	wio sync.Mutex
+
+	// Used by packets.ControlPacket.WriteTo to ensure thread-safe writes
+	sync.Locker
 }
 
 // SetDeadline sets both the read and write deadlines
@@ -141,9 +146,6 @@ func (c *websocketConnector) SetDeadline(t time.Time) error {
 
 // Write writes data to the websocket
 func (c *websocketConnector) Write(p []byte) (int, error) {
-	c.wio.Lock()
-	defer c.wio.Unlock()
-
 	err := c.WriteMessage(websocket.BinaryMessage, p)
 	if err != nil {
 		return 0, err

--- a/packets/packets.go
+++ b/packets/packets.go
@@ -5,6 +5,7 @@ import (
 	"fmt"
 	"io"
 	"net"
+	"sync"
 )
 
 // PacketType is a type alias to byte representing the different
@@ -53,6 +54,20 @@ type (
 		FixedHeader
 	}
 )
+
+// NewThreadSafeConn wraps net.Conn with a mutex. ControlPacket uses it in
+// WriteTo method to ensure parallel writes are thread-Safe.
+func NewThreadSafeConn(c net.Conn) net.Conn {
+	type threadSafeConn struct {
+		net.Conn
+		sync.Locker
+	}
+
+	return &threadSafeConn{
+		Conn:   c,
+		Locker: &sync.Mutex{},
+	}
+}
 
 // WriteTo operates on a FixedHeader and takes the option values and produces
 // the wire format byte that represents these.
@@ -275,6 +290,10 @@ func (c *ControlPacket) WriteTo(w io.Writer) (int64, error) {
 
 	buffers = append(net.Buffers{header.Bytes()}, buffers...)
 
+	if safe, ok := w.(sync.Locker); ok {
+		safe.Lock()
+		defer safe.Unlock()
+	}
 	return buffers.WriteTo(w)
 }
 

--- a/packets/packets_test.go
+++ b/packets/packets_test.go
@@ -3,7 +3,10 @@ package packets
 import (
 	"bufio"
 	"bytes"
+	"io"
+	"net"
 	"reflect"
+	"sync"
 	"testing"
 
 	"github.com/stretchr/testify/assert"
@@ -420,5 +423,60 @@ func BenchmarkPublish_Buffers(b *testing.B) {
 
 	for n := 0; n < b.N; n++ {
 		pp.Buffers()
+	}
+}
+
+func BenchmarkWriteTo(b *testing.B) {
+	r, w := io.Pipe()
+	done := make(chan int, 1)
+
+	go func(r io.Reader, done chan int) {
+		for {
+			_, err := ReadPacket(r)
+			if err != nil {
+				b.Error(err)
+			}
+
+			select {
+			case <-done:
+				return
+			default:
+			}
+		}
+	}(r, done)
+
+	// Wrap Writer with Locker to ensure WriteTo will be thread-safe.
+	type wrap struct {
+		io.Writer
+		sync.Locker
+	}
+	safeWriter := wrap{
+		Writer: w,
+		Locker: &sync.Mutex{},
+	}
+
+	b.RunParallel(func(pb *testing.PB) {
+		for pb.Next() {
+			x := NewControlPacket(PUBLISH)
+			x.Content.(*Publish).QoS = 0
+			x.Content.(*Publish).Topic = "testTopic"
+			x.Content.(*Publish).PacketID = uint16(100)
+			x.Content.(*Publish).Payload = []byte("testPayload")
+			pp := x.Content.(*Publish)
+			for n := 0; n < b.N; n++ {
+				pp.WriteTo(safeWriter)
+			}
+		}
+	})
+
+	done <- 1
+}
+
+func TestNewThreadSafeWrapper(t *testing.T) {
+	var conn net.Conn
+	ts := NewThreadSafeConn(conn)
+
+	if _, ok := ts.(sync.Locker); !ok {
+		t.Error("NewThreadSafeConn does not implement sync.Locker")
 	}
 }

--- a/paho/client.go
+++ b/paho/client.go
@@ -33,7 +33,11 @@ type (
 	// are required to be set, defaults are provided for Persistence, MIDs,
 	// PingHandler, PacketTimeout and Router.
 	ClientConfig struct {
-		ClientID      string
+		ClientID string
+		// Conn is the connection to broker.
+		// BEWARE that most wrapped net.Conn implementations like tls.Conn are
+		// not thread safe for writing. To fix, use packets.NewThreadSafeConn
+		// wrapper or extend the custom net.Conn struct with sync.Locker.
 		Conn          net.Conn
 		MIDs          MIDService
 		AuthHandler   Auther

